From a069c61871dfcdb154a6bb9bcddb92f4607172f2 Mon Sep 17 00:00:00 2001
From: Ray Donnelly <mingw.android@gmail.com>
Date: Sun, 26 Feb 2017 14:33:31 +0000
Subject: [PATCH] [PR 77333] Set clone call fntype to callee type

2017-02-03  Martin Jambor  <mjambor@suse.cz>

        PR target/77333
        * cgraph.c (redirect_call_stmt_to_callee): Set call stmt fntype to
        type of the decl if lhs has compatible type.

Taken from https://gcc.gnu.org/bugzilla/show_bug.cgi?id=77333#c17
---
 gcc/cgraph.c | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/gcc/cgraph.c b/gcc/cgraph.c
index 6ba222b4bdc..621f1beab20 100644
--- a/gcc/cgraph.c
+++ b/gcc/cgraph.c
@@ -1414,7 +1414,15 @@ cgraph_edge::redirect_call_stmt_to_callee (void)
 	new_stmt = chkp_copy_call_skip_bounds (new_stmt);
 
       gimple_call_set_fndecl (new_stmt, e->callee->decl);
-      gimple_call_set_fntype (new_stmt, gimple_call_fntype (e->call_stmt));
+
+      tree lhs = gimple_call_lhs (new_stmt);
+      tree decl_restype = TREE_TYPE (TREE_TYPE (e->callee->decl));
+      if (lhs
+	  && (VOID_TYPE_P (decl_restype)
+	      || !useless_type_conversion_p (TREE_TYPE (lhs), decl_restype)))
+	gimple_call_set_fntype (new_stmt, gimple_call_fntype (e->call_stmt));
+      else
+	gimple_call_set_fntype (new_stmt, TREE_TYPE (e->callee->decl));
 
       if (gimple_vdef (new_stmt)
 	  && TREE_CODE (gimple_vdef (new_stmt)) == SSA_NAME)
-- 
2.11.1

